### DISCUSSION TITLE
Volume Node Affinity Conflict

### DISCUSSION THREAD

**Discussion by brunooliveiramac** [strimzi/strimzi-kafka-operator#8833] [Q&A]
I'm facing an issue when using persistent volumes in google cloud Kubernetes.
Basically, I do have a  cluster with a Kafka node, and this node has 3 zones:
 - us-central1-f
 - us-central1-b
 - us-central1-c
 
 I do have node affinity so my strmzi Kafka clusters are scheduled on the Kafka node.
 The problem is, one of my volumes is being created by strimzi in a different node, other than Kafka kubernetes node, and my kafka cluster is not being able to schedule, complaining about a volume affinity conflict.

![image](https://github.com/strimzi/strimzi-kafka-operator/assets/10716628/c9b8a35a-c813-4097-bb56-882c7cc88b51)

**Comment by scholzj** [comment]
What is the `volumeBindingMode` in your Storage class? Is it `WaitForFirstConsumer`?

**Reply by brunooliveiramac** [reply]
Hey scholzj
Its **immediate**

You mean that we should use a WaitForFirstConsumer, so the binding and provisioning of a PV to a PVC will be delayed until a pod using the PVC is created. This way it would ensure that the PVC/PV is on the correct node?

![image](https://github.com/strimzi/strimzi-kafka-operator/assets/10716628/e7a69ee6-bd93-456c-a9a4-36165ee66d79)

**Reply by scholzj** [reply]
`Immediate` means that when the PVC is created, the storage class creates the PV and binds it. It will use whatever logic it has to pick up the node / zone. At that point, it does not know what will the pod constraints be when it is actually used. So I think you should try to move to `WaitForFirstConsumer` -> which creates the PV and binds it only when the Pod tries to mount it. At that point, it can see the constraints of the Pod and should be able to adjust how the PV is created to them.

So I guess you should either change the Storage Class - or assuming you have it as Immediate for some reason, you can create another Storage CLass with the `WaitForFirstConsumer` mode which you will use for Kafka / ZooKeeper. You can specify which storage class will be used in the storage section in the Kafka CR.

**Reply by brunooliveiramac** [reply]
It worked, thanks a lot [scholzj](https://github.com/scholzj)
